### PR TITLE
add standard for binaries

### PR DESCRIPTION
**Pull request by jdkandersson**
Adds a standard for what to do when needing a binary for one of our projects.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -118,7 +118,21 @@ This ensures that the tests are run on the same Python version as the charm
 will be running in production, catching any issues related to mismatched Python
 versions.
 
-# CI-CD
+## Downloading Binaries
+
+It is a better experience to acquire binaries using an `apt` package or `snap`
+compared to downloading a bianry. Both have features, such as automated updates,
```
Comment (amandahla) [inline CONTRIBUTING.md:124]: s/bianry/binary

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -118,7 +118,21 @@ This ensures that the tests are run on the same Python version as the charm
...
+which enhance the security of the system that makes use of the binaries.
+
+Instead of downloading a binary and using it in our projects, create an
+[`apt`](https://wiki.debian.org/HowToPackageForDebian) package or
+[`snap`](https://snapcraft.io/docs/getting-started). Instead of using a
```
Comment (amandahla) [inline CONTRIBUTING.md:129]: I believe the guidance regarding creating a ROCK instead of using a Docker image could have its own topic, what do you think?

Comment (jdkandersson) [reply CONTRIBUTING.md]: Good point, I removed it in this PR

File: CONTRIBUTING.md (hunk 1 above)
Comment (jdkandersson) [reply CONTRIBUTING.md]: thanks, fixed

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -118,7 +119,19 @@ This ensures that the tests are run on the same Python version as the charm
 will be running in production, catching any issues related to mismatched Python
 versions.
 
-# CI-CD
+## Downloading Binaries
```
Comment (mthaddon) [inline CONTRIBUTING.md:122]: I think the key point about downloading binaries is that we should avoid downloading binaries unless they're from trusted sources (and ideally source that Canonical has control over, so those binaries can be rebuilt with security updates if appropriate). The Ubuntu Archives qualify as a trusted source for downloading binaries. Snaps that are published by "canonical" would also qualify, but we should likely be careful of pretty much anything else. For instance, we wouldn't necessarily want to download a debian package from some random member of the community's PPA, or a snap package from some random individual.

If we can't get a binary from a trusted source what approach should we take? I believe downloading from source, verifying the contents of the download match expectations, and then building the code ourselves is the right approach, but I think this is what we should be clarifying here.

What I'm not clear on is how we treat packages in pypi.

Comment (mthaddon) [reply CONTRIBUTING.md:122]: To me this is still worded to say that it's more about the experience than the security implications. I believe the motivation for this is more from a security perspective, as well as preferring building from source as it gets us a step closer to the SOSS story (i.e. it gives us the possibility of rebuilding code/binaries with security patches applied even if upstream doesn't offer that). Perhaps it's worth some wider discussion to confirm if my assumptions are correct though.

Comment (jdkandersson) [reply CONTRIBUTING.md:122]: What wording changes do you think would address that? I agree that it is both a security and experience issue we're addressing with this standard

**Review by merkata — DISMISSED** [review]

File: CONTRIBUTING.md (hunk 3 above)
Comment (jdkandersson) [reply CONTRIBUTING.md:122]: Thanks for providing the alternate wording, should be updated now

**Review by yanksyoon — APPROVED** [review]
Thanks for the explanation!

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -118,7 +119,37 @@ This ensures that the tests are run on the same Python version as the charm
...
+
+Downloading binaries is only permitted from what are classed as trusted sources.
+These are:
+
+* The Ubuntu Archives (for debian packages)
+* Snaps
+  [owned by the "canonical" account](https://snapcraft.io/publisher/canonical)
+* [PyPi](https://pypi.org/)
+
+These sources are considered trusted because we are confident that we understand
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply CONTRIBUTING.md:132]: Yeh I changed my mind about PyPI being out of scope. For now we can consider PyPI to be trustworthy in terms of generally downloading packages from it. Each package should be reviewed before we add it as a dependency

**Review by cbartz — APPROVED** [review]

**Pull request merged**